### PR TITLE
libbpf-tools/funclatency: support regexp pattern matching kernel func…

### PR DESCRIPTION
**Pull request by chenyuezhou**
Added support for kernel function regular matching

add: 

* string_helper.c/string_helper.h: Provide a string replace operation, such as replacing `*` with `.*` (`vfs_*->vfs_.*`) .
* kprobe_helper.cc/kprobe_helper.h: Provide a way to get all regular matching kernel functions. 

modify:

* Makefile: 
* funclatency.c: Added support for kernel function regular matching.
* trace_helper.c/trace_helper.h: Add `label` field in function `print_log2_hist`.
* other tools: Update the usage of the `print_log2_hist` function.

example

```
$uname -r 
5.10.35
```

```
$./funclatency -m 'vfs_*'
Tracing 70 functions for 'vfs_*'.  Hit Ctrl-C to exit
^C

Function=vfs_open
     msec                : count    distribution
         0 -> 1          : 196      |****************************************|

Function=vfs_readv
     msec                : count    distribution
         0 -> 1          : 3        |****************************************|

Function=vfs_writev
     msec                : count    distribution
         0 -> 1          : 4        |****************************************|

Function=vfs_read
     msec                : count    distribution
         0 -> 1          : 393      |****************************************|

Function=vfs_write
     msec                : count    distribution
         0 -> 1          : 455      |****************************************+|
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 2        |                                        |

Function=vfs_getattr_nosec
     msec                : count    distribution
         0 -> 1          : 161      |****************************************|

Function=vfs_getattr
     msec                : count    distribution
         0 -> 1          : 158      |****************************************|

Function=vfs_statx
     msec                : count    distribution
         0 -> 1          : 6        |****************************************|

Function=vfs_fstat
     msec                : count    distribution
         0 -> 1          : 146      |****************************************|

Function=vfs_fsync_range
     msec                : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 1        |****************************************|

Function=vfs_statfs
     msec                : count    distribution
         0 -> 1          : 40       |****************************************|

Detaching...
```

### CODE REVIEW

**Comment by chenhengqi** [comment]
I wonder if C++ is applicable/acceptable to libbpf-tools.

**Comment by chenyuezhou** [comment]
The reason I use `C++` is that I need some convenient data structures such as `set<>` provided by the `C++ STL`. Of course we could have implemented the data structure in `C`, but that would have brought in additional revew.

If the 'C++' code is not acceptable for 'libbpf-tools', I would also be happy to submit an implementation of C

File: libbpf-tools/trace_helpers.h [hunk 1]
```diff
@@ -53,7 +53,8 @@ partitions__get_by_dev(const struct partitions *partitions, unsigned int dev);
 const struct partition *
 partitions__get_by_name(const struct partitions *partitions, const char *name);
 
-void print_log2_hist(unsigned int *vals, int vals_size, const char *val_type);
+void print_log2_hist(unsigned int *vals, int vals_size, const char *val_type,
```
Comment (chenhengqi) [inline libbpf-tools/trace_helpers.h:56]: Why not add a new API and minimize the changes ?

File: libbpf-tools/kprobe_helpers.cc [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      line = line.substr(pos + 1);
+      kprobe_blacklist_set.insert(line);
+    }
+  }
+
+  size_t count = 0, list_size = (*sz) < 0 ? 0 : (*sz);
+  int in_init_section = 0, in_irq_section = 0;
+  regex pt("^.*\\.cold(\\.\\d+)?$");
+  regex event_re (pattern);
+  ifstream kallsyms_stream { "/proc/kallsyms" };
```
Comment (chenhengqi) [inline libbpf-tools/kprobe_helpers.cc:53]: `trace_helpers.h` already has something like this.

File: libbpf-tools/kprobe_helpers.cc [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+#include <regex>
+#include <fstream>
+#include <sstream>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "kprobe_helpers.h"
+
+using namespace std;
```
Comment (chenhengqi) [inline libbpf-tools/kprobe_helpers.cc:15]: `using namespace std` is not a good practice.

IMO, the impl here is not so idiomatic in C++ :)

**Review by chenhengqi — CHANGES_REQUESTED** [review]

File: libbpf-tools/funclatency.c [hunk 4]
```diff
@@ -6,7 +6,7 @@
  *
  * TODO:
  * - support uprobes on libraries without -p PID. (parse ld.so.cache)
- * - support regexp pattern matching and per-function histograms
+ * - support regexp pattern matching user function and per-function histograms
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/funclatency.c:9]: Hi. Do you think this tool does not need to support regular matching of user-space functions?

File: libbpf-tools/trace_helpers.h (hunk 1 above)
Comment (chenyuezhou) [reply libbpf-tools/trace_helpers.h:56]: I added a print of the label, which I think is necessary, just like the `print_log2_hist` function provided by BCC Python. If I add a new API, there will be a copy of the same code.

File: libbpf-tools/kprobe_helpers.cc (hunk 2 above)
Comment (chenyuezhou) [reply libbpf-tools/kprobe_helpers.cc:53]: Thank you for reminding.

I just read "/proc/kallsyms" and follow my logic for each matching kernel function. The functions provided by 'trace_helpers' do not seem very useful for this.

File: libbpf-tools/Makefile [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
 OUTPUT := .output
 CLANG ?= clang
+C++ ?= c++
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/Makefile]: fixed.

File: libbpf-tools/Makefile [hunk 6]
```diff
@@ -76,16 +82,20 @@ $(OUTPUT) $(OUTPUT)/libbpf:
...
 
 $(patsubst %,$(OUTPUT)/%.o,$(APPS)): %.o: %.skel.h
 
 $(OUTPUT)/%.o: %.c $(wildcard %.h) | $(OUTPUT)
 	$(call msg,CC,$@)
 	$(Q)$(CC) $(CFLAGS) $(INCLUDES) -c $(filter %.c,$^) -o $@
 
+$(OUTPUT)/%.o: %.cc $(wildcard %.h) | $(OUTPUT)
+	$(call msg,C++,$@)
+	$(Q)$(C++) $(CFLAGS) $(INCLUDES) -c $(filter %.cc,$^) -o $@
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/Makefile]: fixed

File: libbpf-tools/kprobe_helpers.cc (hunk 3 above)
Comment (chenyuezhou) [reply libbpf-tools/kprobe_helpers.cc]: fixed

File: libbpf-tools/funclatency.c [hunk 7]
```diff
@@ -250,11 +254,11 @@ static int attach_uprobes(struct funclatency_bpf *obj)
 	return ret;
 }
 
-static int attach_probes(struct funclatency_bpf *obj)
+static int attach_probes(struct funclatency_bpf *obj, char *funcname)
 {
-	if (strchr(env.funcname, ':'))
-		return attach_uprobes(obj);
-	return attach_kprobes(obj);
+  if (strchr(funcname, ':'))
```
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/funclatency.c:259]: My tab was set to 2 spaces.

File: libbpf-tools/funclatency.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenhengqi) [reply libbpf-tools/funclatency.c:9]: I mean the wording.

File: libbpf-tools/funclatency.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (chenhengqi) [reply libbpf-tools/funclatency.c:259]: Just follow the convention.

File: libbpf-tools/trace_helpers.h (hunk 1 above)
Comment (chenhengqi) [reply libbpf-tools/trace_helpers.h:56]: No need for copying.
```
void print_xxx(foo);
void print_xxx_with_label(foo, label);
```
And
```
void print_xxx(foo)
{
    print_xxx_with_label(foo, NULL);
}
```

File: libbpf-tools/kprobe_helpers.cc (hunk 2 above)
Comment (chenhengqi) [reply libbpf-tools/kprobe_helpers.cc]: Why?

File: libbpf-tools/kprobe_helpers.cc (hunk 3 above)
Comment (chenhengqi) [reply libbpf-tools/kprobe_helpers.cc]: I mean the entire file. Especially `char ***`. With C++, we have better option.

File: libbpf-tools/trace_helpers.h (hunk 1 above)
Comment (chenyuezhou) [reply libbpf-tools/trace_helpers.h:56]: Yes, but I need to check that "vals" is empty first, so this code needs a duplicate copy.

File: libbpf-tools/kprobe_helpers.cc (hunk 3 above)
Comment (chenyuezhou) [reply libbpf-tools/kprobe_helpers.cc]: Can you give me some advice? I'm not good at C++.

File: libbpf-tools/kprobe_helpers.cc (hunk 2 above)
Comment (chenyuezhou) [reply libbpf-tools/kprobe_helpers.cc]: 'trace_helper' provides a function to get all kernel function names directly. But I want to process the kernel function name at the same time as getting each kernel function name, which is faster.

File: libbpf-tools/funclatency.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/funclatency.c]: fixed.

File: libbpf-tools/funclatency.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (chenyuezhou) [reply libbpf-tools/funclatency.c]: fixed.

**Comment by chenyuezhou** [comment]
Sry. This PR lacks discussion on related issue, which is a bit abrupt. I will close it.

**Pull request closed without merging**